File: app/src/main/java/za/org/grassroot2/services/Resource.java
package za.org.grassroot2.services;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import static za.org.grassroot2.services.Status.ERROR;
import static za.org.grassroot2.services.Status.LOADING;
import static za.org.grassroot2.services.Status.SERVER_ERROR;
import static za.org.grassroot2.services.Status.SUCCESS;

public class Resource<T> {
    @NonNull public final Status status;
    @Nullable public final T data;
    @Nullable public final String message;

    private Resource(@NonNull Status status, @Nullable T data, @Nullable String message) {
        this.status = status;
        this.data = data;
        this.message = message;
    }

    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(SUCCESS, data, null);
    }

    public static <T> Resource<T> error(String msg, @Nullable T data) {
        return new Resource<>(ERROR, data, msg);
    }

    public static <T> Resource<T> serverError(String msg, @Nullable T data) {
        return new Resource<>(SERVER_ERROR, data, msg);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(LOADING, data, null);
    }
}

File: app/src/main/java/za/org/grassroot2/services/Resource.kt
package za.org.grassroot2.services

import za.org.grassroot2.services.Status.*

class Resource<T> private constructor(val status: Status, val data: T?, val message: String?) {
    companion object {

        fun <T> success(data: T): Resource<T> = Resource(SUCCESS, data, null)
        fun <T> error(msg: String, data: T?) = Resource(ERROR, data, msg)
        fun <T> serverError(msg: String, data: T?)= Resource(SERVER_ERROR, data, msg)
        fun <T> loading(data: T?): Resource<T> = Resource(LOADING, data, null)

    }
}
